src/goose/brokers/redis/cron/registry.clj

```diff
@@ -0,0 +1,64 @@
+(ns goose.brokers.redis.cron.registry
+  (:require [goose.defaults :as d]
+            [goose.brokers.redis.commands :as redis-cmds]
+            [taoensso.carmine :as car]
+            [goose.job :as j]
+            [goose.cron.parsing :as cron-parsing]
+            [goose.utils :as u]))
+
+(defn registry-entry [cron-schedule job-description]
+  {:id              (str (random-uuid))
+   :cron-schedule   cron-schedule
+   :job-description job-description})
+
+(defn register-cron
+  [conn cron-schedule job-description]
+  (let [entry             (registry-entry cron-schedule job-description)
+        next-run-epoch-ms (cron-parsing/next-run-epoch-ms cron-schedule)]
+    (redis-cmds/enqueue-sorted-set conn
+                                   d/prefixed-cron-schedules-queue
+                                   next-run-epoch-ms
+                                   entry)
+    entry))
+
+(defn set-due-time [cron-entry next-run-epoch-ms]
+  (car/zadd d/prefixed-cron-schedules-queue "XX" next-run-epoch-ms cron-entry))
+
+(defn- enqueue-jobs-to-ready-on-priority [jobs]
+  (doseq [[queue-key grouped-jobs] (group-by :prefixed-queue jobs)]
+    (apply car/rpush queue-key grouped-jobs)))
+
+(defn- enqueue-due-cron-entries
+  [due-cron-entries]
+  (let [jobs-to-enqueue          (map (comp j/from-description :job-description) due-cron-entries)
+        entries-by-next-due-time (group-by (comp cron-parsing/next-run-epoch-ms :cron-schedule)
+                                           due-cron-entries)]
+    (enqueue-jobs-to-ready-on-priority jobs-to-enqueue)
+    (doseq [[due-epoch-ms cron-entries] entries-by-next-due-time
+            cron-entry cron-entries]
+      (set-due-time cron-entry due-epoch-ms))))
+
+(defn- due-cron-entries-command []
+  (car/zrangebyscore d/prefixed-cron-schedules-queue
+                     redis-cmds/sorted-set-min
+                     (u/epoch-time-ms)
+                     "limit"
+                     0
+                     d/cron-entries-pop-limit))
+
+(defn due-cron-entries
+  [redis-conn]
+  (not-empty
+    (redis-cmds/wcar* redis-conn
+      (due-cron-entries-command))))
+
+(defn find-and-enqueue-cron-entries
+  "Returns truthy if due cron entries were found."
+  [redis-conn]
+  (redis-cmds/with-transaction redis-conn
+    (car/watch d/prefixed-cron-schedules-queue)
+    (let [due-cron-entries (not-empty (car/with-replies (due-cron-entries-command)))]
+      (car/multi)
+      (when due-cron-entries
```

Comment:
Two commands can be merged into `when-let`

Reply:
This can't be changed, because the `multi` command must be issued after finding the `due-cron-entries`.

Reply:
Following code should give you same results. Please cmiiw
```clojure
    (when-let [due-cron-entries (not-empty (car/with-replies (due-cron-entries-command)))]
      (car/multi)
       (let [jobs-to-enqueue (map (comp j/from-description :job-description) due-cron-entries)]
         (enqueue-jobs-to-ready-on-priority jobs-to-enqueue)
         (doseq [cron-entry due-cron-entries]
           (set-due-time cron-entry))))
```

Reply:
@olttwa here, the problem is that the `multi` command becomes conditional (it is inside the `when-let`). `car/atomic` _always_ issues an `EXEC` command at the end, so the body must contain a `MULTI` in all possible code paths.

In the case that no jobs are found, this code will blow up saying that an `EXEC` was issued without a `MULTI`. I ran into this myself.